### PR TITLE
Add --ignore flag to extractcode

### LINKED ISSUE #1870: Provide an option in `extractcode` to extract sepecific extension files
For instance, I would only want to extract `.rpm`, `.zip`, `.tar.gz`, but not `.jar` files.

The current `extractcode` does not have the ability for user to define which files do the user want to extract or ignore from the extraction.

### PR DESCRIPTION
**Pull request by JRavi2**
Signed-off-by: Ravi <jravi248@gmail.com>

Fixes #1870

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

**Comment by JRavi2** [comment]
@majurg Thanks for the review. 
From what I've understood, the `--ignore` feature in scancode is executed by calling the `is_ignored()` function in commoncode/ignore.py. So in src/extractcode/archive.py/should_extract(), instead of using the `shouldIgnore` flag, should I call the `is_ignored()` function from commoncode, like scancode does?

**Comment by JRavi2** [comment]
@majurg I've made some changes to try to match the behavior with that of scancode. Can you please review them?

Also, I'm not able to figure out:
- Why the Travis CI build is failing
- Why there is such a drastic decrease in coverage in the Codecov report

Can you please help me with these.
Once the changes are approved I'll add the tests in.

**Comment by pombredanne** [comment]
@JRavi2 you wrote:

> I've made some changes to try to match the behavior with that of scancode. Can you please review them?
> 
> Also, I'm not able to figure out:
> 
>     * Why the Travis CI build is failing

That's not your fault...

>> Checking links in thirdparty/wheel-0.33.6-py2.py3-none-any.whl.ABOUT
>>  3. https://bitbucket.org/pypa/wheel 

This URL in the ABOUT file is dead.

>     * Why there is such a drastic decrease in coverage in the Codecov report

 That's a covecov issue

File: src/extractcode/extract.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def extract_files(location, kinds=extractcode.default_kinds, recurse=False):
             if recurse:
                 if TRACE:
                     logger.debug('extract:walk: recursing on target: %(target)r' % locals())
-                for xevent in extract(target, kinds, recurse):
+                for xevent in extract(target, ignored_extensions, kinds, recurse):
```
Comment (pombredanne) [inline src/extractcode/extract.py]: Here and elsewhere I think using named kwargs is usually better and a more robust approach: so it is fine to apply that change as in:
`for xevent in extract(target=target, kinds=kinds, recurse=recurse, ignored_extensions=ignored_extensions):`
both here and elsewhere you made changes

**Comment by JRavi2** [comment]
@pombredanne Thanks for the review. I'll soon add in commits for the suggested changes.

**Comment by JRavi2** [comment]
@pombredanne Sorry for the delay. I've added the suggested changes.

**Comment by pombredanne** [comment]
The test failure on Travis is unrelated and a false alarm, so I will ignore that

File: src/extractcode/archive.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
 from commoncode import functional
 from commoncode.system import on_linux
 from commoncode.system import py2
+from commoncode.ignore import is_ignored
```
Comment (pombredanne) [inline src/extractcode/archive.py]: Sorry to be nit picking but could you keep this import sorted? ignore comes before system

File: src/extractcode/archive.py [hunk 3]
```diff
@@ -103,13 +104,15 @@ def can_extract(location):
...
-def should_extract(location, kinds):
+def should_extract(location, kinds, ignored_extensions=()):
     """
     Return True if this location should be extracted based on the provided
     kinds
     """
     location = os.path.abspath(os.path.expanduser(location))
-    if get_extractor(location, kinds):
+    ignored_extensions = {extension : 'User ignore: Supplied by --ignore' for extension in ignored_extensions}
+    shouldIgnore = is_ignored(location, ignored_extensions)
```
Comment (pombredanne) [inline src/extractcode/archive.py]: We tend to prefer snake_case vs camelCase ... here that would mean `shouldIgnore` should rather be `should_ignore`

File: src/scancode/api.py [hunk 4]
```diff
@@ -312,7 +312,7 @@ def get_file_info(location, **kwargs):
     return result
 
 
-def extract_archives(location, recurse=True, replace_originals=False):
+def extract_archives(location, recurse=True, replace_originals=False, ignored_extensions = ()):
```
Comment (pombredanne) [inline src/scancode/api.py]: Nit picking again, I can spot a few extra spaces in `ignored_extensions = ()` ... may be rather `ignored_extensions=()`

**Comment by JRavi2** [comment]
@pombredanne I need some guidance for the tests. 
I know how to add the tests. I want to know what tests to add.

**Comment by pombredanne** [comment]
@JRavi2 There is two things to look for now:
1. add tests

> Now, if we were to run ./extractcode . --ignore *.tar, it would ignore b.tar and would extract a.zip, b.zip , but will ignore c.tar. Will this test be enough?

That's a start as an integration test but likely not enough. Think about adding also unit tests for each function that you modified to verify that they behave correctly and do what they are supposed to do.

2. are you sure that you are ignoring only extensions? Or rather ignoring glob patterns?

**Comment by pombredanne** [comment]
in anycase you tests in https://github.com/nexB/scancode-toolkit/pull/1946#issuecomment-598773905 in a start alright :+1:

**Comment by JRavi2** [comment]
@pombredanne I am using the `is_ignored` function from commoncode, which I believe uses the glob patters only, so this option would ignore glob patterns only.
Regarding the unit tests, where should the test for `should_extract` function in `archive.py` be added? \
I see that there are different classes to test the extraction of different types of archives, but there is none specifically to test `should_extract`.
I also see that currently there is no existing test for `extract_archives` function in `test_api.py`, so is the test for that required?
Also, thanks for your patience :)

**Comment by JRavi2** [comment]
@pombredanne I've added two tests, please have a look at them. For the rest two functions (`should_extract`, `extract_archives`), I need help as asked in the previous comment.

**Comment by JRavi2** [comment]
@rpotter12 I apologize for the late response. I've resolved the conflicts. Thanks :)

File: src/scancode/extract_cli.py [hunk 5]
```diff
@@ -90,11 +90,12 @@ class ExtractCommand(utils.BaseCommand):
 @click.option('--quiet', is_flag=True, default=False, help='Do not print any summary or progress message.')
 @click.option('--shallow', is_flag=True, default=False, help='Do not extract recursively nested archives (e.g. not archives in archives).')
 @click.option('--replace-originals', is_flag=True, default=False, help='Replace extracted archives by the extracted content.')
+@click.option('--ignore', default=[], multiple=True, help='Ignore files with a particular extension')
```
Comment (rpotter12) [inline src/scancode/extract_cli.py:93]: add full stop at the end of help like others have :)

File: src/extractcode/extract.py [hunk 6]
```diff
@@ -125,7 +125,7 @@ def extract(location, kinds=extractcode.default_kinds, recurse=False, replace_or
     """
     processed_events = []
     processed_events_append = processed_events.append
-    for event in extract_files(location, kinds, recurse):
+    for event in extract_files(location, kinds, recurse, ignored_extensions):
```
Comment (majurg) [inline src/extractcode/extract.py:128]: is `ignored_extensions` always a file extention? What if I do something like `--ignore=specific-file.tar.gz`

File: src/scancode/extract_cli.py (hunk 5 above)
Comment (majurg) [inline src/scancode/extract_cli.py:93]: `--ignore` should work the same way as it does in scancode, meaning you can ignore whole filename or directories, not just file extensions.

File: src/scancode/extract_cli.py [hunk 7]
```diff
@@ -159,8 +160,9 @@ def display_extract_summary():
     extract_result_with_errors = []
     unique_extract_events_with_errors = set()
     has_extract_errors = False
+    
     extractibles = extract_archives(
-        abs_location, recurse=not shallow, replace_originals=replace_originals)
+        abs_location, recurse=not shallow, replace_originals=replace_originals, ignored_extensions=ignore)
```
Comment (majurg) [inline src/scancode/extract_cli.py:165]: Same comment as above. We probably want to keep variable naming consistent. why have the variable `name `ignored_extensions` and not just `ignore`?

File: tests/extractcode/test_extract.py [hunk 8]
```diff
@@ -1112,9 +1112,24 @@ def test_extract_zipslip_tar_posix(self):
 
         warns = [r.warnings for r in result if r.warnings]
         assert [] == warns
-
+
```
Comment (majurg) [inline tests/extractcode/test_extract.py:1115]: Is this white space? if so, we should probably remove this.

File: tests/extractcode/test_extract.py [hunk 9]
```diff
@@ -1112,9 +1112,24 @@ def test_extract_zipslip_tar_posix(self):
...
         assert [] == warns
-
+        
     def test_extract_always_returns_a_generator_and_not_a_list(self):
         # a test for #1996 to ensure that progress is displayed "progressively"
         test_dir = self.get_test_loc('extract/generator', copy=True)
         result = extract.extract(test_dir)
         assert isinstance(result, GeneratorType)
+    
+    def test_extract_ignore(self):
```
Comment (majurg) [inline tests/extractcode/test_extract.py:1122]: We probably want to have more test cases here. We want to make sure `--ignore` can ignore:
1) full filenames
2) directory names
3) various patterns

We should at least have tests cases for all of the above. I would review the current set of extractcode tests and model them after that.

**Comment by majurg** [comment]
@JRavi2 Ping!

**Comment by JRavi2** [comment]
Sorry for the silence! 
This option would also work for ignoring particular files/directories as we are just providing it a glob pattern, so I should change the variable names to a more appropriate one. But, I can't change it to just `ignore` as it would conflict with another ignore being used [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/extractcode/extract.py#L156). Does `ignore_pattern` sound okay? Or, is there another solution to this?

**Comment by majurg** [comment]
@JRavi2 is that particular `ignore` used for anything?

@pombredanne Maybe you have some insight here.

**Comment by JRavi2** [comment]
@majurg I believe this ignore is used to ignore certain default files while walking through the directory. So I just realized that, what we could also do, is just append to the list that it ignores, with the list that the user provides. 
But, the question for the name still remains.

**Comment by agschrei** [comment]
So as announced in #15 I have found the time to run a real-world test on this PR using the [docker-ce git repository](https://github.com/docker/docker-ce).

The following example command is an attempt to cover all three types of arguments the `--ignore` flag should be able to handle

> We want to make sure `--ignore` can ignore:
> 
> 1. full filenames
> 2. directory names
> 3. various patterns

```bash
extractcode --verbose --ignore="*sparse*.tar" --ignore="*issue*.tar" --ignore="/docker-ce/components/engine/vendor/archive/tar/testdata/star.tar"  --ignore="**tarsum/testdata/**" <location of docker-ce clone on disk>
```

Running it behaves exactly as expected and excludes:  

- [x] the absolute path `/docker-ce/components/engine/vendor/archive/tar/testdata/star.tar`  
- [x] any tar archives with `sparse` in their name  
- [x] any tar archives with 'issue' in their name (this exclusion is actually necessary because these archives break libarchive on extraction)  
- [x] any file within the tarsum/testdata subdirectory  

@JRavi2 thank you for your efforts, this contribution is going to be extremely useful for anybody regularly scanning debian package sources.

**Comment by majurg** [comment]
@agschrei Thanks for the feedback! This will be added soon then :)

**Comment by JRavi2** [comment]
@agschrei Glad you found it useful :)
@majurg I'll add the tests in today as soon as possible. Though, I'm still not sure what to rename the variables to?

**Comment by JRavi2** [comment]
@majurg I've added in most of the asked changes. Please review them. :)
The only thing left, if the changes are approved, would be the renaming of the parameters.

**Comment by majurg** [comment]
@JRavi2 I think `ignore_pattern` works as a name, since this is what it actually is.

**Pull request merged**